### PR TITLE
BUG: Consistent division by zero behavior for Index/Series

### LINKED ISSUE #13843: BUG/API Series/Index integer floordiv/mod inconsistency
`Series` looks to be coerced to `float` if the result contains `NaN` or `inf`. Otherwise `Index` doesn't. Make it to be consistent in either way.
#### Code Sample, a copy-pastable example if possible

```
pd.Series([1]) // pd.Series([0])
#0    inf
# dtype: float64

pd.Index([1]) // pd.Index([0])
# Int64Index([0], dtype='int64')

np.array([1]) // np.array([0])
# array([0])
```

```
pd.Series([1]) % pd.Series([0])
#0   NaN
# dtype: float64

pd.Index([1]) % pd.Index([0])
# Int64Index([0], dtype='int64')

np.array([1]) % np.array([0])
# array([0])
```
#### output of `pd.show_versions()`

on current master.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #13843

We have two very-similar functions for masking division by zero for Index vs Series.  The Index version is the more correct of the two, so I'm trying to make the Series one behave more like the Index one, eventually just using the Index one for both.

The hard part here is that IntegerArray, SparseArray, and SparseSeries have their own arithmetic implementations, and we test that they behave the same as Series.  Several attempts to patch these implementations have failed (and motivated #27302, which this is rebased on top of), so for now this PR patches the tests instead of the implementations.

I think the IntegerArray implementation should be relatively easy to patch.  SparseArray I'll need some help (cc @TomAugspurger) with `__setitem__` and/or `putmask`.

### CODE REVIEW

File: pandas/core/sparse/frame.py [hunk 1]
```diff
@@ -74,6 +75,8 @@ def __init__(
         dtype=None,
         copy=False,
     ):
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: didn't we just merge this change?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/sparse/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/sparse/frame.py]: No, this is in #27302.

Comment (jreback) [reply pandas/core/sparse/frame.py]: do you want to move that first?

Comment (jbrockmendel) [reply pandas/core/sparse/frame.py]: This PR is rebased on top of that, on the assumption that will get merged before this

Comment (jreback) [reply pandas/core/sparse/frame.py]: ok

File: pandas/tests/sparse/frame/test_frame.py [hunk 2]
```diff
@@ -424,6 +426,11 @@ def _compare_to_dense(a, b, da, db, op):
             sparse_result = op(a, b)
             dense_result = op(da, db)
 
+            if op in [operator.floordiv, ops.rfloordiv] or isinstance(op, LambdaType):
```
Comment (jreback) [inline pandas/tests/sparse/frame/test_frame.py]: why is this not just ``callable(op)``?

Comment (jbrockmendel) [reply pandas/tests/sparse/frame/test_frame.py]: Because we don't want e.g. `operator.add`

File: pandas/tests/arithmetic/test_numeric.py [hunk 3]
```diff
@@ -332,6 +332,9 @@ def test_ser_divmod_zero(self, dtype1, any_real_dtype):
         right = pd.Series([0, 2]).astype(dtype2)
 
         expected = left // right, left % right
```
Comment (jreback) [inline pandas/tests/arithmetic/test_numeric.py]: in future PR can you give some comments on what you ar doing here (filling np.inf on the integer divide)

File: pandas/tests/sparse/frame/test_frame.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/sparse/frame/test_frame.py]: k can you add that as a comment (future PR ok), as its not obvious

**Comment by jreback** [comment]
k looks fine (except for the comments which can be addressed as a followup), but we need a note for the changes in Series yes? (e.g. they are very edge casey but need a mention)

File: pandas/tests/arithmetic/test_numeric.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_numeric.py]: sure

File: pandas/tests/sparse/frame/test_frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/sparse/frame/test_frame.py]: sure

**Comment by jbrockmendel** [comment]
will add note.  do we want to try to fix the IntegerArray and SparseArray behavior here too?  IntegerArray I think I can figure out but SparseArray I've beat my head against and got nowhere.

**Comment by jreback** [comment]
> will add note. do we want to try to fix the IntegerArray and SparseArray behavior here too? IntegerArray I think I can figure out but SparseArray I've beat my head against and got nowhere.

no let's do as followups.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green (note that I just merged a patch to make the numpy dev branch pass) so you may need to rebase again.

**Comment by jbrockmendel** [comment]
ping

**Pull request merged**